modules/statistics/AnalyticsAdapter.js

```diff
@@ -1,130 +1,297 @@
+
+        if (type !== TYPE_OPERATIONAL && type !== TYPE_PAGE
+            && type !== TYPE_UI && type !== TYPE_TRACK) {
+            logger.error(`Unknown event type: ${type}`);
+
+            return false;
+        }
+
+        if (type === TYPE_PAGE) {
+            return Boolean(event.name);
+        }
+
+        // Try to set some reasonable default values in case some of the
+        // parameters required by the handler API are missing.
+        event.action = event.action || event.name || event.actionSubject;
+        event.actionSubject = event.actionSubject || event.name || event.action;
+        event.source = event.source || event.name || event.action
+            || event.actionSubject;
+
+        if (!event.action || !event.actionSubject || !event.source) {
+            logger.error(
+                'Required field missing (action, actionSubject or source)');
+
+            return false;
+        }
+
+        // Track events have additional required fields.
+        if (type === TYPE_TRACK) {
+            event.objectType = event.objectType || 'generic-object-type';
+            event.containerType
+                = event.containerType || 'generic-container-type';
+
+            if (!event.objectType || !event.objectId
+                || !event.containerType || !event.containerId) {
+                logger.error(
+                    'Required field missing (containerId, containerType, '
+                        + 'objectId or objectType)');
+
+                return false;
+            }
+        }
+
+        return true;
     }
 
     /**
-     * Dispose analytics. Clears all handlers.
+     * Saves an event to the cache, if the cache is enabled.
+     * @param event the event to save.
+     * @returns {boolean} true if the event was saved, and false otherwise (i.e.
+     * if the cache was disabled).
+     * @private
      */
-    dispose() {
-        cacheAnalytics.drainCachedEvents();
-        this.analyticsHandlers.clear();
-        this.disposed = true;
+    _maybeCacheEvent(event) {
+        if (this.cache) {
+            this.cache.push(event);
+
+            // We limit the size of the cache, in case the user fails to ever
+            // set the analytics handlers.
+            if (this.cache.length > MAX_CACHE_SIZE) {
+                this.cache.splice(0, 1);
+            }
+
+            return true;
+        }
+
+        return false;
+
     }
 
     /**
-     * Sets the handlers that are going to be used to send analytics and send
-     * the cached events.
-     * @param {Array} handlers the handlers
+     *
+     * @param event
+     * @private
      */
-    setAnalyticsHandlers(handlers) {
-        if (this.disposed) {
-            return;
+    _sendEvent(event) {
+        if (this._maybeCacheEvent(event)) {
+            // The event was consumed by the cache.
+        } else {
+            // We append the permanent properties at the time we send the event,
+            // not at the time we receive it.
```

Comment:
What is the reasoning again for setting permanent properties on send instead of receive?

Reply:
I wasn't sure, so I preserved the existing behavior. The difference would manifest when the cache is enabled, so before the instance is initialized. So it makes sense to add the permanent properties after they have been initialized.